gcc/rust/expand/rust-macro-builtins-asm.cc

```diff
@@ -16,5 +16,528 @@
 // <http://www.gnu.org/licenses/>.
 
-#include "rust-macro-builtins.h"
-#include "rust-macro-builtins-helpers.h"
+#include "rust-macro-builtins-asm.h"
+
+namespace Rust {
+std::map<AST::InlineAsmOption, std::string> InlineAsmOptionMap{
+  {AST::InlineAsmOption::PURE, "pure"},
+  {AST::InlineAsmOption::NOMEM, "nomem"},
+  {AST::InlineAsmOption::READONLY, "readonly"},
+  {AST::InlineAsmOption::PRESERVES_FLAGS, "preserves_flags"},
+  {AST::InlineAsmOption::NORETURN, "noreturn"},
+  {AST::InlineAsmOption::NOSTACK, "nostack"},
+  {AST::InlineAsmOption::MAY_UNWIND, "may_unwind"},
+  {AST::InlineAsmOption::ATT_SYNTAX, "att_syntax"},
+  {AST::InlineAsmOption::RAW, "raw"},
+};
+
+int
+parseDirSpec (Parser<MacroInvocLexer> &parser, TokenId last_token_id)
+{
+  return 0;
+}
+
+int
+parse_clobber_abi (Parser<MacroInvocLexer> &parser, TokenId last_token_id,
+		   InlineAsmContext &inlineAsmCtx)
+{
+  // clobber_abi := "clobber_abi(" <abi> *("," <abi>) [","] ")"
+  // PARSE EVERYTHING COMMITTEDLY IN THIS FUNCTION, WE CONFIRMED VIA clobber_abi
+  // identifier keyword
+  auto &inlineAsm = inlineAsmCtx.inlineAsm;
+  auto token = parser.peek_current_token ();
+  if (!parser.skip_token (LEFT_PAREN))
+    {
+      // TODO: Raise error exactly like rustc if left parenthesis is not
+      // encountered.
+      token = parser.peek_current_token ();
+
+      // TODO: Error reporting shifted to the left 1 character, I'm not sure
+      // why.
+      if (token->get_id () == last_token_id)
+	{
+	  rust_error_at (parser.peek_current_token ()->get_locus (),
+			 "expected `(`, found end of macro arguments");
+	  return -1;
+	}
+      else
+	{
+	  rust_error_at (
+	    parser.peek_current_token ()->get_locus (),
+	    "expected `(`, found `%s`",
+	    parser.peek_current_token ()->get_token_description ());
+	}
+      return -1;
+    }
+
+  if (parser.skip_token (RIGHT_PAREN))
+    {
+      // TODO: We encountered a "clobber_abi()", which should be illegal?
+      // https://github.com/rust-lang/rust/blob/c00957a3e269219413041a4e3565f33b1f9d0779/compiler/rustc_builtin_macros/src/asm.rs#L381
+      rust_error_at (
+	parser.peek_current_token ()->get_locus (),
+	"at least one abi must be provided as an argument to `clobber_abi`");
+      return -1;
+    }
+
+  std::vector<AST::TupleClobber> new_abis;
+
+  token = parser.peek_current_token ();
+
+  while (token->get_id () != last_token_id && token->get_id () != RIGHT_PAREN)
+    {
+      // Check if it is a string literal or not, codename: <ABI> in ABNF
+      if (token->get_id () == STRING_LITERAL)
+	{
+	  // TODO: Caring for span in here.
+	  new_abis.push_back ({token->as_string (), token->get_locus ()});
+	}
+      else
+	{
+	  // TODO: We encountered something that is not string literal, which
+	  // should be illegal, please emit the correct error
+	  // https://github.com/rust-lang/rust/blob/b92758a9aef1cef7b79e2b72c3d8ba113e547f89/compiler/rustc_builtin_macros/src/asm.rs#L387
```

Comment:
this is good - add an extra `rust_unreachable` here so that the parser doesn't just keep going silently :)
```suggestion
	{
	  // TODO: We encountered something that is not string literal, which
	  // should be illegal, please emit the correct error
	  // https://github.com/rust-lang/rust/blob/b92758a9aef1cef7b79e2b72c3d8ba113e547f89/compiler/rustc_builtin_macros/src/asm.rs#L387
	  rust_unreachable ();
```